tests/test_bookmark.py

```diff
@@ -0,0 +1,83 @@
+            "teams": {"updatedAt": "2020-11-06T20:17:17Z"},
+            "users": {"updatedAt": "2019-10-10T03:23:15.466Z"},
+        }
+    }
+
+    @property
+    def start_date(self):
+        return "2015-03-25T00:00:00Z"
+
+    @staticmethod
+    def name():
+        return "tt_outreach_bookmark_test"
+
+    def streams_to_test(self):
+        # Skip streams due to lack of test data
+        streams_to_exclude = {"mailings", "duties", "sequence_templates"}
+        return self.expected_stream_names().difference(streams_to_exclude)
+
+    def calculate_new_bookmarks(self):
+        """
+        Calculates new bookmarks by looking through sync 1 data to determine a bookmark
+        that will sync 2 records in sync 2 (plus any necessary look back data)
+        """
+        new_bookmarks = {}
+        for stream, records in self.synced_records_1.items():
+            replication_method = self.expected_replication_method(stream)
+            if replication_method == self.INCREMENTAL:
+                look_back = self.expected_lookback_window(stream)
+                replication_key = self.expected_replication_keys(stream)
+                assert len(replication_key) == 1
+                replication_key = next(iter(replication_key))
+
+                # get the replication values that are prior to the lookback window
+                replication_values = sorted({
+                    message['data'][replication_key] for message in records['messages']
+                    if message['action'] == 'upsert'
+                    and self.parse_date(message['data'][replication_key]) <
+                    self.parse_date(self.get_bookmark_value(
+                        self.state_1, self.get_stream_id(stream))) - look_back})
+                print(
+                    f"unique replication values for stream {stream} are: {replication_values}")
+
+                # There should be 3 or more records (prior to the look back window)
+                # so we can set the bookmark to get the last 2 records (+ the look back)
+                # self.assertGreater(len(replication_values), 2,
+                #                    msg="We need to have more than two replication dates "
+                #                        "to test a stream")
+                try:
+                    # Tap expects date format "%Y-%m-%dT%H:%M:%S.000Z" which is not handled by
+                    # base test suite so handling it in the test class explicitly
+                    new_bookmarks[self.get_stream_id(stream)] = {
+                        replication_key:
+                        self.timedelta_formatted(self.parse_date(replication_values[-2]),
+                                                 date_format=self.bookmark_format).replace(".000000Z", ".000Z")}
+                except IndexError as e:
+                    raise Exception(f"{stream}: {replication_values}") from e
+        return new_bookmarks
```

Comment:
The only thing I see different in this method that was overwritten is the `.replace(".000000Z", ".000Z")}` in the last line.  I think you could make it more visible what you are changing from the original method and don't need to repeat all of the logic.  The change for the try catch should be in the tap-tester method if you think it an improvement.

Can you rewrite this something like:

```
    def calculate_new_bookmarks(self):

    new_bookmarks = BookmarkTest.calculate_new_bookmarks()
    new_bookmarks =  {key: {k, v.replace(".000000Z", ".000Z") for val in value} for key, value in new_bookmarks}
```

I'm not 100% sure of that code but something like that

Reply:
Fixed